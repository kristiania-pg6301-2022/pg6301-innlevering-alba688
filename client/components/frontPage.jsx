import React from "react";
import { Link } from "react-router-dom";
import { useLoader } from "../util/useLoader";
import { fetchJSON } from "../util/http";

export function FrontPage() {
  const { data: score, loading } = useLoader(async () =>
    fetchJSON("/api/score")
  );

  return (
    <div>
      <h1>Quiz Application</h1>
      {loading && <div>Loading...</div>}
      {score && (
        <div>
          Score: {score.correct} correct out of {score.answered} questions.
        </div>
      )}
      <br />
      <Link to={"/question"}>
        <button>Answer a new question</button>
      </Link>
    </div>
  );
}
