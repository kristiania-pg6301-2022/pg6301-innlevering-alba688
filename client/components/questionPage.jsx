import React from "react";

import { postJSON, fetchJSON } from "../util/http.jsx";
import { useLoader } from "../util/useLoader";
import { Link } from "react-router-dom";

function QuestionPage({ question, onReload }) {
  async function handleAnswer(answer) {
    const { id } = question;
    postJSON("/api/answer", { id, answer });
    onReload();
  }

  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((a) => question.answers[a])
        .map((a) => (
          <p key={a}>
            <button onClick={() => handleAnswer(a)}>
              {question.answers[a]}
            </button>
          </p>
        ))}
      <Link to="/">Go back</Link>
    </div>
  );
}

export function QuestionComponent() {
  const { loading, error, reload, data } = useLoader(async () =>
    fetchJSON("/api/question")
  );

  const question = data;

  if (loading) {
    return <h2>Loading...</h2>;
  }

  if (error) {
    return (
      <div style={{ border: "1px solid red", background: "Pink" }}>
        An error occurred: {error.toString()}
      </div>
    );
  }

  return <QuestionPage question={question} onReload={reload} />;
}
