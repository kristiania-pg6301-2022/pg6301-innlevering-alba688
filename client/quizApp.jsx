import React from "react";
import { Route, Routes } from "react-router-dom";
import { FrontPage } from "./components/frontPage";
import { QuestionComponent } from "./components/questionPage";

export function Application() {
  return (
    <Routes>
      <Route path={"/"} element={<FrontPage />} />
      <Route path={"/question"} element={<QuestionComponent />} />
      <Route path={"*"} element={<h1>Not found</h1>} />
    </Routes>
  );
}
