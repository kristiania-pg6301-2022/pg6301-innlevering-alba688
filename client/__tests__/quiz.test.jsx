import * as React from "react";
import ReactDOM from "react-dom";
import { Application } from "../quizApp.jsx";
import { FrontPage } from "../components/frontPage";
import { QuestionComponent } from "../components/questionPage";
import { MemoryRouter } from "react-router";

const testQuestion = {
  id: 201,
  question: "What year is it?",
  answers: {
    answer_a: "1988",
    answer_b: "1673",
    answer_c: "2022",
    answer_d: "3121",
  },
};

describe("Client Quiz Application", () => {
  it("shows a random question", async function () {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/question"]}>
        <QuestionComponent />
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
  });
});
