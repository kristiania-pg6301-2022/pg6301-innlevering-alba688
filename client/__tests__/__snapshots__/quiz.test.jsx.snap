// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Quiz Application shows a random question 1`] = `"<h2>Loading...</h2>"`;
